firefox-bootstrap/bootstrap.js

```diff
@@ -0,0 +1,259 @@
+          "nsISupportsWeakReference"]),
+        onStateChange: function(webProgress, request, flag, status) {
+          if (flag & Ci.nsIWebProgressListener.STATE_STOP) {
+            if (request.originalURI && request.originalURI.host === WEBCOMPAT_DOMAIN) {
+              let mm = browser.selectedBrowser.messageManager;
+              browser.removeProgressListener(this);
+
+              // did we successfully get a screenshot?
+              if (tabData && tabData.canvasData) {
+                mm.loadFrameScript(FRAMESCRIPT, true);
+                mm.sendAsyncMessage(SCREENSHOT_MESSAGE, {
+                  screenshot: tabData.canvasData
+                });
+
+                mm.addMessageListener(ACK_MESSAGE, function ackFn(message) {
+                  mm.removeMessageListener(ACK_MESSAGE, ackFn);
+                  resolve();
+                });
+              } else {
+                resolve();
+              }
+              httpRequestObserver.removeObserver();
+            }
+          }
+        }
+      };
+
+      httpRequestObserver.addObserver();
+      browser.addProgressListener(loadedListener);
+      browser.loadOneTab(
+        `https://${WEBCOMPAT_DOMAIN}/issues/new?url=${encodeURIComponent(tabData.url)}`,
+        {inBackground: false});
+    });
+  },
+
+  getScreenshot(browser) {
+    const FRAMESCRIPT = `chrome://webcompat-reporter/content/tab-frame.js?${Date.now()}`;
+    const TABDATA_MESSAGE = "WebCompat:SendTabData";
+
+    return new Promise((resolve) => {
+      let mm = browser.selectedBrowser.messageManager;
+
+      mm.loadFrameScript(FRAMESCRIPT, true);
+      mm.addMessageListener(TABDATA_MESSAGE, function receiveFn(message) {
+        mm.removeMessageListener(TABDATA_MESSAGE, receiveFn);
+        resolve([browser, message.json]);
+      });
+    });
+  },
+
+  reportIssue(xulDoc) {
+    Promise.resolve(xulDoc.defaultView.gBrowser).then(this.getScreenshot)
+                                                .then(this.openWebCompatTab)
+                                                .catch(Cu.reportError);
+  }
+};
+
+// use enabled pref as a way for tests (e.g. test_contextmenu.html) to disable
+// the addon when running.
+let prefObserver = function(aSubject, aTopic, aData) {
+  let enabled = Services.prefs.getBoolPref(PREF_WC_REPORTER_ENABLED);
+  if (enabled) {
+    WebCompatReporter.init();
+  } else {
+    WebCompatReporter.uninit();
+  }
+};
+
+function startup(aData, aReason) { // eslint-disable-line no-unused-vars
+  // Observe pref changes and enable/disable as necessary.
+  Services.prefs.addObserver(PREF_WC_REPORTER_ENABLED, prefObserver, false);
+
+  // temporary for add-on port, set this in all.js somewhere, probably
+  Services.prefs.setBoolPref(PREF_WC_REPORTER_ENABLED, true);
+  // Only initialize if pref is enabled.
+  let enabled = Services.prefs.getBoolPref(PREF_WC_REPORTER_ENABLED);
+  if (enabled) {
+    WebCompatReporter.init();
```

Comment:
So, I've built an XPI and tested the startup of this addon, and 💣

```
Exception running bootstrap method startup on webcompat-reporter@mozilla.com: TypeError: Services.wm.getMostRecentWindow(...) is null (resource://gre/modules/addons/XPIProvider.jsm -> jar:file:///Users/dennis/Work/.firefox-profiles/default-2qJxGQ/extensions/webcompat-reporter@mozilla.com.xpi!/bootstrap.js:118:19)
JS Stack trace: init@resource://gre/modules/addons/XPIProvider.jsm -> jar:file:///Users/dennis/Work/.firefox-profiles/default-2qJxGQ/extensions/webcompat-reporter@bootstrap.js:118:19)
```

I'm not too surprised by this. We do a lot of stuff in parallel these days, and it's quite possible to have the startup method running before the window was created.

I wonder... given you already had issues with multiple windows and you've had to add the somewhat *hacky* listener attribute on the browser, maybe there is an alternate approach.

What about replacing this logic with something like

```js
class WindowListener {
  onOpenWindow() {
    let win = window.QueryInterface(Ci.nsIInterfaceRequestor).
      getInterface(Ci.nsIDOMWindow);

    win.addEventListener("load", function onLoad() {
      win.removeEventListener("load", onLoad, false);

      // initialize a new WebCompatReporter instance here, store somewhere
      // probably within the window
    });
  }
}

function startup(aData, aReason) {
  // ...
  Services.wm.addListener(new WindowListener());
  // ...
}
```

I have not tested this approach at all and wrote this code just to start a discussion. This approach would mean we'd have a dedicated WebCompatReporter instance per window, which is a little bit of memory overhead, but I don't think it's too bad. Maybe there's a better approach. So far, I've used the [SDKs `startup()` method wrapper](https://dxr.mozilla.org/mozilla-central/rev/13f49da109ea460665ad27c8497cb1489548450c/addon-sdk/source/lib/sdk/addon/bootstrap.js#90) whenever I've dealt with multiple windows, so I don't have any experience in that particular case. It could be wise to read through the SDKs `startup()` implementation, or just ask someone for feedback.